### PR TITLE
Refactor Commands, Parsers, and Person Fields to Improve Validation, Defensive Coding, and Test Coverage

### LINKED ISSUE #177: [Enhancement] Improve Defensive Coding, Parser Validation, and Test Heuristic Coverage
## **Description**

Several components in the codebase lack consistent defensive coding, validation handling, and systematic test design.

This results in:

* Inconsistent null handling (missing `requireNonNull` / undocumented assumptions)
* Partial validation logic (e.g., duplicate prefix handling not applied uniformly)
* Test cases lacking clear structure based on testing heuristics
* Misalignment between implementation and test expectations

---

## **Scope of Changes**

This issue affects multiple areas of the codebase, including:

* Command logic (e.g., `AddCommand`)
* Parser logic (e.g., `AddCommandParser`)
* UI components (e.g., `PersonCard`)
* Associated unit and parser test files

Due to the cross-cutting nature of these improvements, changes span multiple files (~40+), but all are related to the same objective of improving code quality and test reliability.

---

## **Proposed Solution**

* Add defensive checks (`requireNonNull`, `assert`) where appropriate
* Standardise validation flow across parsers
* Ensure Javadoc reflects actual assumptions and behaviour
* Refactor tests to follow:

  * Equivalence Partitioning (EP)
  * Boundary Value Analysis (BVA)
  * Multiple input heuristics (e.g., test invalid inputs individually before combining)
* Align test expectations with actual implementation logic

---

## **Expected Outcome**

* Improved code robustness and consistency
* Reduced risk of hidden bugs due to invalid inputs
* Clearer and more systematic test coverage
* Better maintainability across affected components

### PR DESCRIPTION
**Pull request by savagexr3**
## **Description**

The current implementation across multiple commands (`add`, `edit`, `findcom`, `sortcom`), their parsers, and person-related fields (e.g., `Name`, `Phone`, `Email`, `Address`, `Company`, `Remark`, `Tag`) contains inconsistencies in validation, defensive coding, and test design.

These issues reduce code reliability, clarity, and alignment with testing standards.

---

### **1. Defensive Coding Improvements**

* Added `requireNonNull` and `assert` where appropriate
* Ensured assumptions (e.g., non-null inputs) are explicitly documented in Javadoc
* Improved fail-fast behaviour for invalid inputs

---

### **2. Validation Standardisation**

* Unified validation approach across:

  * `AddCommand`, `EditCommand`, `FindComCommand`, `SortComCommand`
  * Corresponding parsers
* Improved handling of:

  * Duplicate prefixes
  * Optional fields (`company`, `remark`, `link`)
* Ensured consistency between:

  * Field validation methods (`getValidationError`)
  * Parser behaviour

---

### **3. Person Field Enhancements**

* Reviewed and aligned validation logic for:

  * `Name`, `Phone`, `Email`, `Address`
  * `Company`, `Remark`, `Tag`
* Ensured constraints (e.g., format, length, spacing) are consistently enforced
* Updated Javadoc to accurately describe validation rules and assumptions

---

### **4. Test Refactoring (Heuristic-Based)**

Refactored unit and parser tests to follow structured testing strategies:

* **Equivalence Partitioning (EP)**

  * Valid vs invalid vs null inputs
* **Boundary Value Analysis (BVA)**

  * Length limits and edge cases
* **Multiple Input Heuristics**

  * Each invalid input tested individually before combining
  * Combined invalid cases added after individual validation

Added inline comments in test cases to clearly indicate:

* Which heuristic is being applied
* Why the test case exists

---

### **5. Code Quality Improvements**

* Increased **cohesion** by ensuring each class/method has a clear responsibility
* Reduced **coupling** by removing redundant logic and improving separation of concerns
* Fixed inconsistencies between implementation and test expectations
* Cleaned up UI-related logic (`PersonCard`) for clarity

---

## **Scope**

This is a cross-cutting refactor affecting multiple components:

* Commands: `AddCommand`, `EditCommand`, `FindComCommand`, `SortComCommand`
* Parsers for the above commands
* Person-related model classes (`Name`, `Phone`, `Email`, `Address`, `Company`, `Remark`, `Tag`)
* Corresponding unit, parser, and integration tests

Changes span multiple files (~40+), but are all aligned under improving validation, defensive coding, and test quality.

---

## **Expected Outcome**

* More robust and consistent validation across all commands and fields
* Improved maintainability through better cohesion and reduced coupling
* Clear, systematic test coverage aligned with CS2103 testing standards
---

## **Issues**
* Closes #177

### CODE REVIEW

File: src/main/java/seedu/clinkedin/logic/parser/ArgumentMultimap.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ public String getPreamble() {
     }
 
     /**
-     * Throws a {@code ParseException} if any of the prefixes given in {@code prefixes} appeared more than
+     * Throws a {@codae ParseException} if any of the prefixes given in {@code prefixes} appeared more than
```
Comment (FacelessSnowman) [inline src/main/java/seedu/clinkedin/logic/parser/ArgumentMultimap.java]: typo here

**Review by patrickstar1337 — APPROVED** [review]
Had a look through everything and it seems okay. Might have missed a few things though.

Maybe consider breaking up the changes into different PRs next time? Not sure if it's a lot because of another PR.

Cheers

**Comment by savagexr3** [comment]
> Had a look through everything and it seems okay. Might have missed a few things though.
> 
> Maybe consider breaking up the changes into different PRs next time? Not sure if it's a lot because of another PR.
> 
> Cheers

Thanks for taking a look!
Yeah this PR got a bit big since I bundled a few related changes together 😅 I’ll break them into smaller PRs next time to make things easier to review.
Thanks again!
